src/segment/filter/blocked_bloom/mod.rs

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) 2024-present, fjall-rs
+// This source code is licensed under both the Apache 2.0 and MIT License
+// (found in the LICENSE-* files in the repository)
+
+mod builder;
+use super::{bit_array::BitArrayReader, AMQFilter, CACHE_LINE_BYTES};
+use crate::{
+    coding::{DecodeError, Encode, EncodeError},
+    file::MAGIC_BYTES,
+};
+pub use builder::Builder;
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use std::io::{Read, Write};
+
+/// Two hashes that are used for double hashing
+pub type CompositeHash = (u64, u64);
+
+pub struct BlockedBloomFilter {
+    /// Raw bytes exposed as bit array
+    inner: BitArrayReader,
+
+    /// Number of hash functions
+    k: usize,
+
+    /// Number of blocks in the blocked bloom filter
+    num_blocks: usize,
+}
+
+impl AMQFilter for BlockedBloomFilter {
+    fn bytes(&self) -> &[u8] {
+        self.inner.bytes()
+    }
+
+    /// Size of bloom filter in bytes
+    #[must_use]
+    fn len(&self) -> usize {
+        self.inner.bytes().len()
+    }
+
+    /// Returns `true` if the hash may be contained.
+    ///
+    /// Will never have a false negative.
+    #[must_use]
+    fn contains_hash(&self, (mut h1, mut h2): CompositeHash) -> bool {
+        let block_idx = h1 % (self.num_blocks as u64);
+
+        for i in 1..(self.k as u64) {
+            h1 = h1.wrapping_add(h2);
+            h2 = h2.wrapping_add(i);
+
+            let idx = h1 % (CACHE_LINE_BYTES as u64);
+
+            // NOTE: should be in bounds because of modulo
+            #[allow(clippy::expect_used, clippy::cast_possible_truncation)]
+            if !self.has_bit(block_idx as usize, idx as usize) {
+                return false;
+            }
+        }
+
+        true
+    }
+
+    /// Returns `true` if the item may be contained.
+    ///
+    /// Will never have a false negative.
+    #[must_use]
+    fn contains(&self, key: &[u8]) -> bool {
+        self.contains_hash(Self::get_hash(key))
+    }
+
+    fn filter_type(&self) -> super::FilterType {
+        super::FilterType::BlockedBloom
+    }
+}
+
+impl Encode for BlockedBloomFilter {
+    fn encode_into<W: Write>(&self, writer: &mut W) -> Result<(), EncodeError> {
+        // Write header
+        writer.write_all(&MAGIC_BYTES)?;
+
+        writer.write_u8(super::FilterType::BlockedBloom as u8)?;
+
+        // NOTE: Hash type (unused)
+        writer.write_u8(0)?;
+
+        writer.write_u64::<LittleEndian>(self.num_blocks as u64)?;
```

Comment:
there's an underlying assumption here that each cache line block will always be 64 bytes

Reply:
Nothing that can be done about that I think.

Most systems use 512 bits; Apple Silicon seems to use 1024 bits, meaning 2 blocks would share a cache line.